src/core/server/plugins/plugin.ts

```diff
@@ -155,8 +155,8 @@ export interface Plugin<
 export type PluginInitializer<
   TSetup,
   TStart,
-  TPluginsSetup extends Record<PluginName, unknown> = {},
-  TPluginsStart extends Record<PluginName, unknown> = {}
+  TPluginsSetup extends object = object,
+  TPluginsStart extends object = object
```

Comment:
What if we change this to `TPluginsSetup extends object = Record<string, unknown>`? Wouldn't that force the consumer to specify a type rather than defaulting to a type whose values are `any`?

Reply:
in what cases it could help us?
1 if a plugin implements `Plugin` interface, it still has to declare all types explicitly, because TS [cannot infer contextual types from an interface](https://github.com/Microsoft/TypeScript/issues/1373)
2 if an object implements `Plugin` interface we have the same problem as [discussed here](https://github.com/elastic/kibana/pull/39329#discussion_r297767388). end users have to declare all required members when we fall back to the `object` type.

There is a [sandbox](https://typescript-play.js.org/#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAlCBAeygBMAeAKGWQBUBlCMTAB2QF5kA3I4UgGmp16YOFDAduvAUNoYcoAM6NmbCAA9IIUouREARgCtCEzgWJlyisFFDZ+yTCADWIIgHcQAPkE05WXBBlUXFkDS0dPSMTSXMSCmtbEHtHFzdPL0ovZABvIUUmVgAKCwgALmQ4EABPBxYApQr-BSCVVgBKJra2AB80KCIAW2AC8gZCli8AbnyQsBKScsqauoagpvlA4LEwTuE55D7UAeHRhjnp2aIWAH4ivZ4+GYBfSkoNFhIJBHQ4RV1NqAAILIYCDeoQQYQcAAtZxMi5ITEII2TAIMAke65V40FiYfToYAIZAFVQLKBLKq1ZCkCAsRTtbFCPEEokkubkykrGl0hlM3H4wnE6zXLE5V6vSjI6zIeotIEVQEgeGkSR5Gg0UnFUqrFoM9Uaw1yrYAOjgM0Nz18GusO05uq27QNho1xqUJv0Fo1EreoEgsEQKCVAHlouiqH5upJHjJIwdODHrc0tlHwtDIgZjOjJJmTEmldtQmntLpc9nOGWwFlEZqJpyKlSHY06AXuntxqpDv0hiMIGNupdNRzSg3uW71i21oXdl0DkcTr3+xcvSLbvcKjGXm8Pl9kD8-rCWgAhUHg9CQ6FgQ+BUNZiQG6Wo9GYxni5mCtlalj15bU2n018cVlD9hWHRZRz-XlAPfVlQNFaDJUfWU1iPRU1lvGJOGdEk6x1ZC9SdIQXXw01zSI5ArXI21xHtEilEI4jXSnD0vRoCUgA) with both examples

Reply:
Ah I was mistaken on how the raw `object` type worked. I thought it would default to essentially `Record<string, any>`. This LGTM 👍